prog/weaver/metrics.go

```diff
@@ -0,0 +1,143 @@
+package main
+
+import (
+	"net/http"
+	"os"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	"github.com/weaveworks/weave/ipam"
+	"github.com/weaveworks/weave/nameserver"
+	"github.com/weaveworks/weave/net/address"
+	weave "github.com/weaveworks/weave/router"
+)
+
+func metricsHandler(router *weave.NetworkRouter, allocator *ipam.Allocator, ns *nameserver.Nameserver) http.Handler {
+	reg := prometheus.NewRegistry()
+	reg.MustRegister(prometheus.NewProcessCollector(os.Getpid(), ""))
+	reg.MustRegister(newMetrics(router, allocator, ns))
+	return promhttp.HandlerFor(reg, promhttp.HandlerOpts{})
+}
+
+type metrics struct {
+	router              *weave.NetworkRouter
+	allocator           *ipam.Allocator
+	ns                  *nameserver.Nameserver
+	connectionCountDesc *prometheus.Desc
```

Comment:
Have you considered storing the descriptors in a list of structs which one field would be responsible for returning associated metric values?